exchanges/request/request_test.go

```diff
@@ -559,77 +513,61 @@ func TestBasicLimiter(t *testing.T) {
-	r, err := New("TestRequest", new(http.Client), WithLimiter(NewBasicRateLimit(time.Minute, 1, 1)))
-	if err != nil {
-		t.Fatal(err)
-	}
-	ctx := context.Background()
+	rl := NewBasicRateLimit(50*time.Millisecond, 1, 1)
+	r, err := New("TestRequest", http.DefaultClient, WithLimiter(rl))
+	require.NoError(t, err, "New requester must not error")
 
-	var resp any
-	err = r.SendPayload(ctx, Auth, func() (*Item, error) {
-		return &Item{
-			Method: http.MethodGet,
-			Path:   testURL,
-			Result: &resp,
-		}, nil
-	}, AuthenticatedRequest)
-	if err != nil {
-		t.Fatal(err)
+	sendIt := func() error {
+		return r.SendPayload(t.Context(), Auth, func() (*Item, error) {
+			return &Item{
+				Method: http.MethodGet,
+				Path:   testURL,
+				Result: new(any),
+			}, nil
+		}, AuthenticatedRequest)
 	}
 
-	err = r.EnableRateLimiter()
-	if err == nil {
-		t.Fatal("error cannot be nil")
-	}
+	t.Run("allow initial request", func(t *testing.T) {
+		require.NoError(t, sendIt(), "sendIt must not error")
+	})
 
-	err = r.DisableRateLimiter()
-	if err != nil {
-		t.Fatal(err)
-	}
+	t.Run("error on redundant enable", func(t *testing.T) {
+		err := r.EnableRateLimiter()
+		assert.ErrorIs(t, err, ErrRateLimiterAlreadyEnabled)
+	})
 
-	err = r.SendPayload(ctx, Auth, func() (*Item, error) {
-		return &Item{
-			Method: http.MethodGet,
-			Path:   testURL,
-			Result: &resp,
-		}, nil
-	}, AuthenticatedRequest)
-	if err != nil {
-		t.Fatal(err)
-	}
+	t.Run("error on redundant disable", func(t *testing.T) {
+		require.NoError(t, r.DisableRateLimiter(), "DisableRateLimiter must not error")
+		err := r.DisableRateLimiter()
+		assert.ErrorIs(t, err, ErrRateLimiterAlreadyDisabled)
+	})
 
-	err = r.DisableRateLimiter()
-	if err == nil {
-		t.Fatal("error cannot be nil")
-	}
+	t.Run("allow requests when disabled", func(t *testing.T) {
+		require.NoError(t, sendIt(), "sendIt must not error")
+	})
 
-	err = r.EnableRateLimiter()
-	if err != nil {
-		t.Fatal(err)
-	}
+	t.Run("allow when re-enabled", func(t *testing.T) {
+		require.NoError(t, r.EnableRateLimiter(), "EnableRateLimiter must succeed")
+		require.NoError(t, sendIt(), "sendIt must not error")
+	})
 
-	ti := time.NewTicker(time.Second)
-	c := make(chan struct{})
-	go func(c chan struct{}) {
-		err = r.SendPayload(ctx, Auth, func() (*Item, error) {
-			return &Item{
-				Method: http.MethodGet,
-				Path:   testURL,
-				Result: &resp,
-			}, nil
-		}, AuthenticatedRequest)
-		if err != nil {
-			log.Fatal(err)
+	t.Run("block excess requests", func(t *testing.T) {
+		// consume the one token
+		require.NoError(t, sendIt(), "sendIt must not error")
+
+		// fire a second request and expect it to block for at least 20ms
+		errCh := make(chan error, 1)
+		go func() {
+			errCh <- sendIt()
+		}()
+
+		select {
+		case err := <-errCh:
+			assert.NoError(t, err, "Rate limiter should block until request is done")
+		case <-time.After(20 * time.Millisecond):
+			// still blocked 🌞
```

Comment:
🚧 I think we need to yield to that goro
🐒 I think it's probably cleaner to use `assert.Never` here
```suggestion
		runtime.Gosched()
		assert.Never(t, func() bool { return len(errCh) > 0 }, 20*time.Millisecond, time.Millisecond, "Rate limiter should block until first request is done")
```

Reply:
Nice alternative.